mmeds/spawn.py

```diff
@@ -162,28 +106,98 @@ def __init__(self, queue, parent_pid, testing=False):
         """
         self.testing = testing
         self.q = queue
-        self.processes = read_processes()
+        self.processes = [] # read_processes()
+        self.running_processes = []
+        self.pipe = pipe
         self.parent_pid = parent_pid
         self.started = []
         super().__init__()
 
     def add_process(self, ptype, process):
         """ Add an analysis process to the list of processes. """
-        self.processes[ptype].append(process)
-        atexit.unregister(write_processes)
-        atexit.register(write_processes, self.processes)
-        atexit.unregister(killall)
-        atexit.register(killall, self.started)
+        error_log('Add process {}, type: {}'.format(process, ptype))
+        #self.processes[ptype].append(process)
+        self.running_processes.append(process)
+        self.write_running_processes()
+
+    def check_processes(self):
+        still_running = []
+        # For each type of process 'upload', 'analysis', 'etc'
+        # Check each process is still alive
+        while self.running_processes:
+            process = self.running_processes.pop()
+            if process.is_alive():
+                still_running.append(process)
+            else:
+                self.processes.append(process)
+                # Send the exitcode of the process
+                self.pipe.send(process.exitcode)
+        for process in still_running:
+            self.running_processes.append(process)
+
+    def read_processes():
+        """
+        Function for reading process access codes back from the log file.
+        Part of the functionality for continuing unfinished analyses on server restart.
+        """
+        if fig.CURRENT_PROCESSES.exists():
+            with open(fig.CURRENT_PROCESSES, 'r') as f:
+                processes = yaml.safe_load(f, Loader=yaml.Loader)
+            for key in processes.keys():
+                for process in processes[key]:
+                    process['is_alive'] = False
+        else:
+            processes = defaultdict(list)
```

Comment:
I think this is a good example of a situation where I might consider adding logging like:
logger.warn("Current_Processes not found, returning defaultdict instead")
Not sure if that's needful here, but I like keeping an eye out for places where an unexpected code path can cause issues further downstream.

Reply:
I should remove this whole function. I have decided against loading general processes from a log. To continue tracking jobs when restarting on Minerva I think I will add something else, like a function that runs 'bjobs' and grabs the output or something

Reply:
Makes sense, you could try setting up checkpointing and automating running jobs from a checkpoint.

Reply:
That is something I should have setup for larger datasets in the future but would address the issue of continuing to monitor an analysis after a mmeds server restart